File: app/lib/utils/transcript_hash.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';
import 'package:omi/backend/schema/transcript_segment.dart';

/// Client copy of `omi:backend/utils/conversations/transcript_hash.py` encoding v5.
///
/// The digest is a wire contract. Bump [transcriptHashEncodingVersion] when
/// the framing changes. The version is not mixed into the digest bytes.
const int transcriptHashEncodingVersion = 5;

const String defaultSpeaker = 'SPEAKER_00';
const int defaultSpeakerId = 0;
const String isUserToken = 'Y';
const String notUserToken = 'N';

class CanonicalTranscriptSegment {
  const CanonicalTranscriptSegment({
    required this.speaker,
    required this.speakerId,
    required this.isUser,
    required this.personId,
    required this.text,
  });

  final String speaker;
  final int speakerId;
  final bool isUser;
  final String? personId;
  final String text;
}

String _stripOrDefault(String? raw, [String defaultValue = '']) {
  if (raw == null) return defaultValue;
  final stripped = raw.trim();
  if (stripped.isEmpty) return defaultValue;
  return stripped;
}

int derivedSpeakerId(String canonicalSpeaker) {
  final parts = canonicalSpeaker.split('_');
  if (parts.length < 2) return defaultSpeakerId;
  return int.tryParse(parts[1]) ?? defaultSpeakerId;
}

CanonicalTranscriptSegment canonicalizeSegment({
  String? speaker,
  int? speakerId,
  bool? isUser,
  String? personId,
  String? text,
}) {
  final canonicalSpeaker = _stripOrDefault(speaker, defaultSpeaker);
  final canonicalPerson = _stripOrDefault(personId);
  return CanonicalTranscriptSegment(
    speaker: canonicalSpeaker,
    speakerId: speakerId ?? derivedSpeakerId(canonicalSpeaker),
    isUser: isUser == true,
    personId: canonicalPerson.isEmpty ? null : canonicalPerson,
    text: _stripOrDefault(text),
  );
}

CanonicalTranscriptSegment canonicalizeTranscriptSegment(TranscriptSegment segment) {
  return canonicalizeSegment(
    speaker: segment.speaker,
    speakerId: segment.speakerId,
    isUser: segment.isUser,
    personId: segment.personId,
    text: segment.text,
  );
}

List<int> _frame(String value) {
  final encoded = utf8.encode(value);
  return [...ascii.encode('${encoded.length}\n'), ...encoded];
}

List<int> canonicalTranscriptBytes(Iterable<CanonicalTranscriptSegment> parts) {
  final out = <int>[];
  for (final part in parts) {
    out
      ..addAll(_frame(part.speaker))
      ..addAll(_frame('${part.speakerId}'))
      ..addAll(_frame(part.isUser ? isUserToken : notUserToken))
      ..addAll(_frame(part.personId ?? ''))
      ..addAll(_frame(part.text));
  }
  return out;
}

String transcriptSha256FromCanonical(Iterable<CanonicalTranscriptSegment> parts) {
  return sha256.convert(canonicalTranscriptBytes(parts)).toString();
}

String transcriptSha256(Iterable<TranscriptSegment> segments) {
  return transcriptSha256FromCanonical(segments.map(canonicalizeTranscriptSegment));
}

String transcriptSha256FromMaps(Iterable<Map<String, Object?>> segments) {
  return transcriptSha256FromCanonical(
    segments.map(
      (segment) => canonicalizeSegment(
        speaker: segment['speaker'] as String?,
        speakerId: segment['speaker_id'] as int?,
        isUser: segment['is_user'] as bool?,
        personId: segment['person_id'] as String?,
        text: segment['text'] as String?,
      ),
    ),
  );
}

File: desktop/macos/Desktop/Sources/LocalInference/TranscriptHash.swift
import CryptoKit
import Foundation

/// Client copy of `backend/utils/conversations/transcript_hash.py` encoding v5.
///
/// The digest is a wire contract with the server projection binder. The version
/// names this framing; it is not mixed into the digest bytes.
enum TranscriptHash {
  static let encodingVersion = 5
  static let defaultSpeaker = "SPEAKER_00"
  static let defaultSpeakerId = 0
  static let isUserToken = "Y"
  static let notUserToken = "N"

  struct Segment: Sendable, Equatable {
    var speaker: String?
    var speakerId: Int?
    var isUser: Bool?
    var personId: String?
    var text: String

    init(
      speaker: String? = nil,
      speakerId: Int? = nil,
      isUser: Bool? = nil,
      personId: String? = nil,
      text: String
    ) {
      self.speaker = speaker
      self.speakerId = speakerId
      self.isUser = isUser
      self.personId = personId
      self.text = text
    }
  }

  struct CanonicalSegment: Sendable, Equatable {
    var speaker: String
    var speakerId: Int
    var isUser: Bool
    var personId: String?
    var text: String
  }

  static func canonicalize(_ segment: Segment) -> CanonicalSegment {
    let speaker = stripOrDefault(segment.speaker, defaultSpeaker)
    let person = stripOrDefault(segment.personId, "")
    return CanonicalSegment(
      speaker: speaker,
      speakerId: segment.speakerId ?? derivedSpeakerId(speaker),
      isUser: segment.isUser == true,
      personId: person.isEmpty ? nil : person,
      text: stripOrDefault(segment.text, "")
    )
  }

  static func sha256(segments: [Segment]) -> String {
    sha256(canonical: segments.map(canonicalize))
  }

  static func sha256(canonical parts: [CanonicalSegment]) -> String {
    let digest = SHA256.hash(data: Data(canonicalBytes(parts)))
    return digest.map { String(format: "%02x", $0) }.joined()
  }

  /// Python `split('_', 1)[1]` — only the first underscore is the cut.
  static func derivedSpeakerId(_ speaker: String) -> Int {
    guard let idx = speaker.firstIndex(of: "_") else { return defaultSpeakerId }
    let rest = speaker[speaker.index(after: idx)...]
    return Int(rest) ?? defaultSpeakerId
  }

  static func canonicalBytes(_ parts: [CanonicalSegment]) -> [UInt8] {
    var out: [UInt8] = []
    out.reserveCapacity(parts.count * 24)
    for part in parts {
      out.append(contentsOf: frame(part.speaker))
      out.append(contentsOf: frame(String(part.speakerId)))
      out.append(contentsOf: frame(part.isUser ? isUserToken : notUserToken))
      out.append(contentsOf: frame(part.personId ?? ""))
      out.append(contentsOf: frame(part.text))
    }
    return out
  }

  private static func stripOrDefault(_ raw: String?, _ defaultValue: String) -> String {
    guard let raw else { return defaultValue }
    let stripped = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    return stripped.isEmpty ? defaultValue : stripped
  }

  private static func frame(_ value: String) -> [UInt8] {
    let encoded = Array(value.utf8)
    return Array("\(encoded.count)\n".utf8) + encoded
  }
}

extension TranscriptionSegmentRecord {
  var hashSegment: TranscriptHash.Segment {
    TranscriptHash.Segment(
      speaker: speakerLabel,
      speakerId: speaker,
      isUser: isUser,
      personId: personId,
      text: text
    )
  }
}
